shared/p2p/service.go

```diff
@@ -119,7 +119,7 @@ func (s *Server) RegisterTopic(topic string, message proto.Message, adapters ...
 			}
 
 			var h Handler = func(ctx context.Context, pMsg Message) {
-				s.emit(feed, msg, msgType)
+				s.emit(pMsg.Ctx, feed, msg, msgType)
 			}
 
 			pMsg := Message{}
```

Comment:
Let's assign a context here and remove the context parameter from the adapter functions, as you suggested.